dwave/optimization/src/c++23/cartesian_product.cpp

```diff
@@ -0,0 +1,43 @@
+// Copyright 2024 D-Wave
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include "dwave-optimization/c++23/cartesian_product.hpp"
+
+#include <numeric>
+#include <iostream>
+
+namespace dwave::optimization {
+
+// Based on https://stackoverflow.com/a/31169617
+// Since std::cartesian_product is only specified for c++23 we
+// add our own implementation of it.
+std::vector<std::vector<ssize_t>> cartesian_product(std::vector<std::vector<ssize_t>> v) {
```

Comment:
It looks like we don't mutate `v`? So probably this should accept a reference so we don't do a copy.

Reply:
Yeah, reason why I removed the reference was because I thought about having the option to pass rvalues. Can definitely have it as a reference in this case